plugins/device/cmd/nvidia/fingerprint.go

```diff
@@ -0,0 +1,244 @@
+	MemoryMiBAttr          = "memory_mib"
+	PowerWAttr             = "power_w"
+	BAR1MiBAttr            = "bar1_mib"
+	DriverVersionAttr      = "driver_version"
+	CoresClockMHzAttr      = "cores_clock_mhz"
+	MemoryClockMHzAttr     = "memory_clock_mhz"
+	PCIBandwidthMBPerSAttr = "pci_bandwidth_mb/s"
+	DisplayStateAttr       = "display_state"
+	PersistenceModeAttr    = "persistence_mode"
+)
+
+// fingerprint is the long running goroutine that detects hardware
+func (d *NvidiaDevice) fingerprint(ctx context.Context, devices chan<- *device.FingerprintResponse) {
+	defer close(devices)
+
+	if d.nvmlClientInitializationError != nil {
+		d.logger.Error("Exiting fingerprinting due to problems with NVML loading")
+		// write empty fingerprint response to let server know that there are
+		// no working Nvidia GPU units
+		devices <- device.NewFingerprint()
+		return
+	}
+
+	// Create a timer that will fire immediately for the first detection
+	ticker := time.NewTimer(0)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			ticker.Reset(d.fingerprintPeriod)
+		}
+		d.writeFingerprintToChannel(devices)
+	}
+}
+
+// writeFingerprintToChannel makes nvml call and writes response to channel
+func (d *NvidiaDevice) writeFingerprintToChannel(devices chan<- *device.FingerprintResponse) {
+	d.logger.Trace("Issuing gonvml request to get fingerprint data for nvidia devices")
+	fingerprintData, err := d.nvmlClient.GetFingerprintData()
+
+	if err != nil {
+		d.logger.Error("Failed to get fingerprint nvidia devices with error", err)
+		devices <- device.NewFingerprintError(err)
+		return
+	}
+
+	d.logger.Trace("Fingerprint query to nvml was successful")
+	// ignore devices from fingerprint output
+	fingerprintDevices := ignoreFingerprintedDevices(fingerprintData.Devices, d.ignoredGPUIDs)
+	// check if any device health was updated or any device was added to host
+	updated := d.checkFingerprintUpdates(fingerprintDevices)
+
+	if updated {
+		commonAttributes := map[string]string{
+			DriverVersionAttr: fingerprintData.DriverVersion,
+		}
+
+		// Group all FingerprintDevices by DeviceName attribute
+		deviceListByDeviceName := make(map[string][]*nvml.FingerprintDeviceData)
+		for _, device := range fingerprintDevices {
+			deviceName := device.DeviceName
+			if deviceName == nil {
+				// nvml driver was not able to detect device name. This kind
+				// of devices are placed to single group with 'notAvailable' name
+				notAvailableCopy := notAvailable
+				deviceName = &notAvailableCopy
+			}
+
+			deviceListByDeviceName[*deviceName] = append(deviceListByDeviceName[*deviceName], device)
+		}
+
+		// Build Fingerprint response with computed groups and send it over the channel
+		deviceGroups := make([]*device.DeviceGroup, len(deviceListByDeviceName))
+		groupCounter := 0
+		for groupName, devices := range deviceListByDeviceName {
+			deviceGroups[groupCounter] = deviceGroupFromFingerprintData(groupName, devices, commonAttributes)
```

Comment:
`deviceGroups = append(deviceGroups, deviceGroupFromFingerprintData(groupName, devices, commonAttributes))`

This coupled with the above change to make means you can delete groupCounter, is easier to read and more idiomatic